server/src/test/java/org/elasticsearch/search/suggest/SuggestionTests.java

```diff
@@ -196,17 +198,20 @@ public void testToXContent() throws IOException {
             suggestion.addTerm(entry);
             BytesReference xContent = toXContent(suggestion, XContentType.JSON, params, randomBoolean());
             assertEquals(
-                    "{\"phrase#suggestionName\":[{"
-                            + "\"text\":\"entryText\","
-                            + "\"offset\":42,"
-                            + "\"length\":313,"
-                            + "\"options\":[{"
-                                + "\"text\":\"someText\","
-                                + "\"highlighted\":\"somethingHighlighted\","
-                                + "\"score\":1.3,"
-                                + "\"collate_match\":true}]"
-                            + "}]"
-                    + "}", xContent.utf8ToString());
+                // @formatter:off
```

Comment:
We have lines like this in a lot of tests. Is there anything we can do to improve this without resorting to manually turning off the formatter?

Reply:
There are options, depending on the code in question. 

### JSON snippets

For this example, we could include whitespace:

```java
String suggestionString = "{"
    + "\"unknownType#suggestionName\": ["
    + "  {"
    + "    \"text\": \"entryText\","
    + "    \"offset\": 42,"
    + "    \"length\": 313,"
    + "    \"options\": [{"
    + "      \"text\": \"someText\","
    + "      \"highlighted\": \"somethingHighlighted\","
    + "      \"score\": 1.3,"
    + "      \"collate_match\": true"
    + "    }]"
    + "  }]"
    + "}";
```

With Java 13, we could use [text blocks](https://openjdk.java.net/jeps/355):

```
String suggestionString = """
    {
    "unknownType#suggestionName": [
      {
        "text": "entryText",
        "offset": 42,
        "length": 313,
        "options": [{
          "text": "someText",
          "highlighted": "somethingHighlighted",
          "score": 1.3,
          "collate_match": true
        }]
      }]
    }
    """;
```

We could also make these snippets more readable by using an `ObjectMapper` with some lenient settings, e.g:

```java
public static String asJson(final String rawJson) throws IOException {
    ObjectMapper mapper = new ObjectMapper();

    mapper.configure(JsonParser.Feature.ALLOW_UNQUOTED_FIELD_NAMES, true);
    mapper.configure(JsonParser.Feature.ALLOW_SINGLE_QUOTES, true);

    final JsonNode jsonNode = mapper.readTree(rawJson);

    return mapper.writeValueAsString(jsonNode);
}
```

Then we can write the JSON like so:

```java
String suggestionString = asJson("{"
    + "'unknownType#suggestionName': ["
    + "  {"
    + "    text: 'entryText',"
    + "    offset: 42,"
    + "    length: 313,"
    + "    options: [{"
    + "      text: 'someText',"
    + "      highlighted: 'somethingHighlighted',"
    + "      score: 1.3,"
    + "      collate_match: true"
    + "    }]"
    + "  }]"
    + "}");
```

Depend how wild we're feeling, we could define these snippets in YAML, and convert to JSON, e.g.:

```java
String suggestionYaml = yamlToJson(
        "unknownType#suggestionName:\n"
        + "  - text: entryText\n"
        + "    offset: 42\n"
        + "    length: 313\n"
        + "    options:\n"
        + "      - text: someText\n"
        + "        highlighted: somethingHighlighted\n"
        + "        score: 1.3\n"
        + "        collate_match: true\n");
```

Though I expect that'll look better with text blocks, once they're available:

```
String suggestionYaml = yamlToJson("""
    unknownType#suggestionName:
      - text: entryText
        offset: 42
        length: 313
        options:
          - text: someText
            highlighted: somethingHighlighted
            score: 1.3
            collate_match: true
    """);
```

### Aligned code

The changes to [SslConfigurationLoader.java](https://github.com/elastic/elasticsearch/pull/47299/files#diff-e479ea9b1c908fc4e30392ff15c413c1) are a case where strings and comments are arranged in meaningful rows. I think this is a case where just exempting the code is the best approach.

On that subject, however, the Painless module is interesting because it has code that has vertical alignment. For example:

   - [DefMath.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/DefMath.java#L1083)
   - [MethodWriter.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/MethodWriter.java#L184)
   - [Operation.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/Operation.java#L30)
   - [WriterConstants.java](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/WriterConstants.java#L137)

I can't decide whether we should turn off the formatter for these. What do you think?

Also, Spotless doesn't appear to have a way to exempt files from formatting, so it formats all the source that ANTLR generates too. Can't do much about that.

Reply:
@rjernst do you want me to do anything here? I could go looking for more code whose format we want to preserve?

Reply:
Long term I think we will want to use text blocks. I wouldn't change to yaml, since we are testing json. Ideally these would actually be converted to xcontentbuilder, since that has earlier error checking (eg invalid json would get caught when running the code, rather than in the request parsing the json). My concern here is starting this precedent of arbitrarily turning off the formatter. Using column aligned strings seems like the best option for now, but we should probably change those in separate PRs from any enabling auto formatting, so that the formatting PRs can remain completely automatic and mechanical in nature, thus simplifying reviewing.